### PR TITLE
[Merged by Bors] - feat(Mathlib/Tactic): implement `infer_opt_param`

### PR DESCRIPTION
**Pull request by urkud**

### CODE REVIEW

**Comment by urkud** [comment]
I don't understand why the `push` CI fails. Any ideas?

**Comment by arthurpaulino** [comment]
No idea why it failed but re-running it did the trick

File: test/InferParam.lean [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+namespace InferParamTest
+
+theorem zero_le_add (a : Nat) (ha : 0 ≤ a := Nat.zero_le a) :
+  0 ≤ a + a :=
+calc
+  0 ≤ a := ha
+  _ ≤ a + a := Nat.le_add_left _ _
+
+-- example : 0 ≤ 2 + 2 := by
+  -- infer_opt_param -- fails
```
Comment (digama0) [inline test/InferParam.lean:12]:
```suggestion
example : 0 ≤ 2 + 2 := by
  fail_if_success infer_opt_param
  decide
```

File: Mathlib/Tactic/InferParam.lean [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+open Lean Elab Tactic Meta
+
+/-- Close a goal of the form `optParam α a` by using `a`. -/
+elab (name := inferOptParam) "infer_opt_param" : tactic => do
+  withMainContext do
+    let tgt_expr ← getMainTarget
+    let goal ← getMainGoal
+    match tgt_expr.getAppFnArgs with
+    | (``optParam, #[_ty, val]) => assignExprMVar goal val; replaceMainGoal []
+    | _ => throwError "`infer_opt_param` only solves goals of the form `optParam _ _`, not {tgt_expr}"
```
Comment (digama0) [inline Mathlib/Tactic/InferParam.lean:31]:
```suggestion
elab (name := inferOptParam) "infer_opt_param" : tactic =>
  liftMetaTactic fun goal => do
    let tgt_expr ← getMVarType goal
    match tgt_expr.getAppFnArgs with
    | (``optParam, #[_ty, val]) => assignExprMVar goal val; pure []
    | _ => throwError
      "`infer_opt_param` only solves goals of the form `optParam _ _`, not {tgt_expr}"
```

File: test/InferParam.lean (hunk 1 above)
Comment (urkud) [reply test/InferParam.lean:12]: Do we already have `fail_if_success`?

File: Mathlib/Tactic/InferParam.lean (hunk 2 above)
Comment (urkud) [reply Mathlib/Tactic/InferParam.lean:31]: Tomorrow I'll ask you about these edits.

Comment (semorrison) [reply Mathlib/Tactic/InferParam.lean:31]: Perhaps we could go over this one in a "public" session? Mario has suggested using `liftMetaTactic` on several different PRs now.

File: test/InferParam.lean (hunk 1 above)
Comment (semorrison) [reply test/InferParam.lean:12]: Yes.

**Pull request closed without merging**